skops/io/_visualize.py

```diff
@@ -0,0 +1,326 @@
+    this function doesn't need to know how nodes can be nested and how safety of
+    a node is determined.
+
+    Parameters
+    ----------
+    node: :class:`skops.io._audit.Node`
+        The current node to visit. Children are visited recursively.
+
+    show: "all" or "untrusted" or "trusted"
+        Whether to print all nodes, only untrusted nodes, or only trusted nodes.
+
+    node_name: str (default="root")
+        The key to the current node. If "key_types" is encountered, it is
+        skipped.
+
+    level: int (default=0)
+        The current level of nesting.
+
+    Yields
+    ------
+    :class:`~NodeInfo`:
+        A dataclass containing the aforementioned information.
+
+    """
+    # key_types is not helpful, as it is artificially added by skops to
+    # circumvent the fact that json only allows keys to be strings. It is not
+    # useful to the user and adds a lot of noise, thus skip key_types.
+    # TODO: check that no funny business is going on in key types
+    if node_name == "key_types":
+        return
+
+    # COMPOSITE TYPES: CHECK ALL ITEMS
+    if isinstance(node, dict):
+        for key, val in node.items():
+            yield from walk_tree(
+                val,
+                node_name=key,
+                level=level,
+            )
+        return
+
+    if isinstance(node, (list, tuple)):
+        # shouldn't be tuple, but check just to be sure
+        for val in node:
+            yield from walk_tree(
+                val,
+                node_name=node_name,
+                level=level,
+            )
+        return
+
+    # NO MATCH: RAISE ERROR
+    if not isinstance(node, Node):
+        raise TypeError(
+            f"Cannot deal with {type(node)}, please report the issue here "
+            "https://github.com/skops-dev/skops/issues"
+        )
+
+    # YIELDING THE ACTUAL NODE INFORMATION HERE
+
+    # Note: calling node.is_safe() on all nodes is potentially wasteful because
+    # it is already a recursive call, i.e. child nodes will be checked many
+    # times. A solution to this would be to add caching to its call.
+    yield NodeInfo(
+        level=level,
+        key=node_name,
+        val=node.format(),
+        is_self_safe=node.is_self_safe(),
+        is_safe=node.is_safe(),
+    )
+
+    # TYPES WHOSE CHILDREN IT MAKES NO SENSE TO VISIT
+    # TODO: For better security, we should check the schema if we return early,
+    # otherwise something nefarious could be hidden inside.
+    if isinstance(node, (NdArrayNode, SparseMatrixNode)) and (node.type != "json"):
+        return
+
+    if isinstance(node, (NdArrayNode, SparseMatrixNode)) and (node.type == "json"):
+        return
+
+    if isinstance(node, FunctionNode):
+        return
+
+    if isinstance(node, JsonNode):
+        pass
+
+    # RECURSE
+    yield from walk_tree(
+        node.children,
+        node_name=node_name,
+        level=level + 1,
+    )
+
+
+def visualize_tree(
```

Comment:
I don't think it matters that it's a _tree_. What about just `visualize`? Or `inspect` (although there's the `inspect` module)?